File: dbflow-processor/src/main/java/com/raizlabs/android/dbflow/processor/definition/method/DatabaseDefinition.java
package com.raizlabs.android.dbflow.processor.definition.method;

import com.raizlabs.android.dbflow.annotation.ConflictAction;
import com.raizlabs.android.dbflow.annotation.Database;
import com.raizlabs.android.dbflow.processor.ClassNames;
import com.raizlabs.android.dbflow.processor.definition.BaseDefinition;
import com.raizlabs.android.dbflow.processor.definition.MigrationDefinition;
import com.raizlabs.android.dbflow.processor.definition.ModelViewDefinition;
import com.raizlabs.android.dbflow.processor.definition.QueryModelDefinition;
import com.raizlabs.android.dbflow.processor.definition.TableDefinition;
import com.raizlabs.android.dbflow.processor.definition.TypeDefinition;
import com.raizlabs.android.dbflow.processor.handler.DatabaseHandler;
import com.raizlabs.android.dbflow.processor.model.ProcessorManager;
import com.raizlabs.android.dbflow.processor.utils.StringUtils;
import com.raizlabs.android.dbflow.processor.validator.ModelViewValidator;
import com.raizlabs.android.dbflow.processor.validator.TableValidator;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;

/**
 * Description: Writes {@link com.raizlabs.android.dbflow.annotation.Database} definitions,
 * which contain {@link com.raizlabs.android.dbflow.annotation.Table},
 * {@link com.raizlabs.android.dbflow.annotation.ModelView}, and {@link com.raizlabs.android.dbflow.annotation.Migration}
 */
public class DatabaseDefinition extends BaseDefinition implements TypeDefinition {

    public String databaseName;

    public int databaseVersion;

    boolean foreignKeysSupported;

    boolean consistencyChecksEnabled;

    boolean backupEnabled;

    public ConflictAction insertConflict;

    public ConflictAction updateConflict;

    public String classSeparator;
    public String fieldRefSeparator; // safe field for javapoet

    public boolean isInMemory;

    private DatabaseHolderDefinition holderDefinition;

    public DatabaseDefinition(ProcessorManager manager, Element element) {
        super(element, manager);
        setPackageName(ClassNames.FLOW_MANAGER_PACKAGE);

        Database database = element.getAnnotation(Database.class);
        if (database != null) {
            databaseName = database.name();
            if (databaseName == null || databaseName.isEmpty()) {
                databaseName = element.getSimpleName().toString();
            }
            if (!isValidDatabaseName(databaseName)) {
                throw new Error("Database name [ " + databaseName + " ] is not valid. It must pass [A-Za-z_$]+[a-zA-Z0-9_$]* " +
                        "regex so it can't start with a number or contain any special character except '$'. Especially a dot character is not allowed!");
            }

            consistencyChecksEnabled = database.consistencyCheckEnabled();
            backupEnabled = database.backupEnabled();

            classSeparator = database.generatedClassSeparator();

            if (!StringUtils.isNullOrEmpty(classSeparator)) {
                // all are $
                fieldRefSeparator = classSeparator;
            } else {
                fieldRefSeparator = classSeparator;
            }

            setOutputClassName(databaseName + classSeparator + "Database");

            databaseVersion = database.version();
            foreignKeysSupported = database.foreignKeysSupported();

            insertConflict = database.insertConflict();
            updateConflict = database.updateConflict();
            isInMemory = database.inMemory();
        }
    }

    public DatabaseHolderDefinition getHolderDefinition() {
        return holderDefinition;
    }

    public void setHolderDefinition(DatabaseHolderDefinition holderDefinition) {
        this.holderDefinition = holderDefinition;
    }

    @Override
    protected TypeName getExtendsClass() {
        return ClassNames.BASE_DATABASE_DEFINITION_CLASSNAME;
    }

    @Override
    public void onWriteDefinition(TypeSpec.Builder typeBuilder) {
        writeConstructor(typeBuilder);
        writeGetters(typeBuilder);
    }

    public void validateAndPrepareToWrite() {
        prepareDefinitions();
        validateDefinitions();
    }

    private void validateDefinitions() {
        // TODO: validate them here before preparing them
        Map<TypeName, TableDefinition> map = new HashMap<>();
        TableValidator tableValidator = new TableValidator();
        for (TableDefinition tableDefinition : getManager().getTableDefinitions(getElementClassName())) {
            if (tableValidator.validate(ProcessorManager.getManager(), tableDefinition)) {
                map.put(tableDefinition.getElementClassName(), tableDefinition);
            }
        }
        getManager().setTableDefinitions(map, getElementClassName());

        Map<TypeName, ModelViewDefinition> modelViewDefinitionMap = new HashMap<>();
        ModelViewValidator modelViewValidator = new ModelViewValidator();
        for (ModelViewDefinition modelViewDefinition : getManager().getModelViewDefinitions(getElementClassName())) {
            if (modelViewValidator.validate(ProcessorManager.getManager(), modelViewDefinition)) {
                modelViewDefinitionMap.put(modelViewDefinition.getElementClassName(), modelViewDefinition);
            }
        }
        getManager().setModelViewDefinitions(modelViewDefinitionMap, getElementClassName());

    }

    private void prepareDefinitions() {
        for (TableDefinition tableDefinition : getManager().getTableDefinitions(getElementClassName())) {
            tableDefinition.prepareForWrite();
        }

        for (ModelViewDefinition modelViewDefinition : getManager().getModelViewDefinitions(getElementClassName())) {
            modelViewDefinition.prepareForWrite();
        }

        for (QueryModelDefinition queryModelDefinition : getManager().getQueryModelDefinitions(getElementClassName())) {
            queryModelDefinition.prepareForWrite();
        }
    }

    private void writeConstructor(TypeSpec.Builder builder) {

        MethodSpec.Builder constructor = MethodSpec.constructorBuilder()
                .addModifiers(Modifier.PUBLIC)
                .addParameter(ClassNames.DATABASE_HOLDER, "holder");

        for (TableDefinition tableDefinition : getManager().getTableDefinitions(getElementClassName())) {
            constructor.addStatement("holder.putDatabaseForTable($T.class, this)", tableDefinition.getElementClassName());
        }

        for (ModelViewDefinition modelViewDefinition : getManager().getModelViewDefinitions(getElementClassName())) {
            constructor.addStatement("holder.putDatabaseForTable($T.class, this)", modelViewDefinition.getElementClassName());
        }

        for (QueryModelDefinition queryModelDefinition : getManager().getQueryModelDefinitions(getElementClassName())) {
            constructor.addStatement("holder.putDatabaseForTable($T.class, this)", queryModelDefinition.getElementClassName());
        }

        Map<Integer, List<MigrationDefinition>> migrationDefinitionMap = getManager().getMigrationsForDatabase(getElementClassName());
        if (migrationDefinitionMap != null && !migrationDefinitionMap.isEmpty()) {
            List<Integer> versionSet = new ArrayList<>(migrationDefinitionMap.keySet());
            Collections.sort(versionSet);
            for (Integer version : versionSet) {
                List<MigrationDefinition> migrationDefinitions = migrationDefinitionMap.get(version);
                Collections.sort(migrationDefinitions, new Comparator<MigrationDefinition>() {
                    @Override
                    public int compare(MigrationDefinition o1, MigrationDefinition o2) {
                        return Integer.valueOf(o2.priority).compareTo(o1.priority);
                    }
                });
                constructor.addStatement("$T migrations$L = new $T()", ParameterizedTypeName.get(ClassName.get(List.class), ClassNames.MIGRATION),
                        version, ParameterizedTypeName.get(ArrayList.class));
                constructor.addStatement("$L.put($L, migrations$L)", DatabaseHandler.MIGRATION_FIELD_NAME,
                        version, version);
                for (MigrationDefinition migrationDefinition : migrationDefinitions) {
                    constructor.addStatement("migrations$L.add(new $T$L)", version, migrationDefinition.getElementClassName(),
                            migrationDefinition.getConstructorName());
                }
            }
        }

        for (TableDefinition tableDefinition : getManager().getTableDefinitions(getElementClassName())) {
            constructor.addStatement("$L.add($T.class)", DatabaseHandler.MODEL_FIELD_NAME, tableDefinition.getElementClassName());
            constructor.addStatement("$L.put($S, $T.class)", DatabaseHandler.MODEL_NAME_MAP, tableDefinition.getTableName(), tableDefinition.getElementClassName());
            constructor.addStatement("$L.put($T.class, new $T(holder, this))", DatabaseHandler.MODEL_ADAPTER_MAP_FIELD_NAME,
                    tableDefinition.getElementClassName(), tableDefinition.getOutputClassName());
        }

        for (ModelViewDefinition modelViewDefinition : getManager().getModelViewDefinitions(getElementClassName())) {
            constructor.addStatement("$L.add($T.class)", DatabaseHandler.MODEL_VIEW_FIELD_NAME, modelViewDefinition.getElementClassName());
            constructor.addStatement("$L.put($T.class, new $T(holder, this))", DatabaseHandler.MODEL_VIEW_ADAPTER_MAP_FIELD_NAME,
                    modelViewDefinition.getElementClassName(), modelViewDefinition.getOutputClassName());
        }

        for (QueryModelDefinition queryModelDefinition : getManager().getQueryModelDefinitions(getElementClassName())) {
            constructor.addStatement("$L.put($T.class, new $T(holder, this))", DatabaseHandler.QUERY_MODEL_ADAPTER_MAP_FIELD_NAME,
                    queryModelDefinition.getElementClassName(), queryModelDefinition.getOutputClassName());
        }

        builder.addMethod(constructor.build());
    }

    private void writeGetters(TypeSpec.Builder typeBuilder) {

        typeBuilder.addMethod(MethodSpec.methodBuilder("getAssociatedDatabaseClassFile")
                .addAnnotation(Override.class)
                .addModifiers(DatabaseHandler.METHOD_MODIFIERS)
                .addStatement("return $T.class", getElementTypeName())
                .returns(ParameterizedTypeName.get(Class.class)).build());

        typeBuilder.addMethod(MethodSpec.methodBuilder("isForeignKeysSupported")
                .addAnnotation(Override.class)
                .addModifiers(DatabaseHandler.METHOD_MODIFIERS)
                .addStatement("return $L", foreignKeysSupported)
                .returns(TypeName.BOOLEAN).build());

        typeBuilder.addMethod(MethodSpec.methodBuilder("isInMemory")
                .addAnnotation(Override.class)
                .addModifiers(DatabaseHandler.METHOD_MODIFIERS)
                .addStatement("return $L", isInMemory)
                .returns(TypeName.BOOLEAN).build());

        typeBuilder.addMethod(MethodSpec.methodBuilder("backupEnabled")
                .addAnnotation(Override.class)
                .addModifiers(DatabaseHandler.METHOD_MODIFIERS)
                .addStatement("return $L", backupEnabled)
                .returns(TypeName.BOOLEAN).build());

        typeBuilder.addMethod(MethodSpec.methodBuilder("areConsistencyChecksEnabled")
                .addAnnotation(Override.class)
                .addModifiers(DatabaseHandler.METHOD_MODIFIERS)
                .addStatement("return $L", consistencyChecksEnabled)
                .returns(TypeName.BOOLEAN).build());

        typeBuilder.addMethod(MethodSpec.methodBuilder("getDatabaseVersion")
                .addAnnotation(Override.class)
                .addModifiers(DatabaseHandler.METHOD_MODIFIERS)
                .addStatement("return $L", databaseVersion)
                .returns(TypeName.INT).build());

        typeBuilder.addMethod(MethodSpec.methodBuilder("getDatabaseName")
                .addAnnotation(Override.class)
                .addModifiers(DatabaseHandler.METHOD_MODIFIERS)
                .addStatement("return $S", databaseName)
                .returns(ClassName.get(String.class)).build());
    }

    /**
     * <p>Checks if databaseName is valid. It will check if databaseName matches regex pattern
     * [A-Za-z_$]+[a-zA-Z0-9_$]*</p> Examples: <ul> <li>database - valid</li> <li>DbFlow1 - valid</li> <li>database.db -
     * invalid (contains a dot)</li> <li>1database - invalid (starts with a number)</li> </ul>
     *
     * @param databaseName database name to validate.
     * @return {@code true} if parameter is a valid database name, {@code false} otherwise.
     */
    private static boolean isValidDatabaseName(final String databaseName) {
        final Pattern javaClassNamePattern = Pattern.compile("[A-Za-z_$]+[a-zA-Z0-9_$]*");
        return javaClassNamePattern.matcher(databaseName).matches();
    }
}

File: dbflow-processor/src/main/java/com/raizlabs/android/dbflow/processor/definition/DatabaseDefinition.kt
package com.raizlabs.android.dbflow.processor.definition

import com.grosner.kpoet.L
import com.grosner.kpoet.S
import com.grosner.kpoet.`return`
import com.grosner.kpoet.constructor
import com.grosner.kpoet.final
import com.grosner.kpoet.modifiers
import com.grosner.kpoet.param
import com.grosner.kpoet.public
import com.grosner.kpoet.statement
import com.raizlabs.android.dbflow.annotation.ConflictAction
import com.raizlabs.android.dbflow.annotation.Database
import com.raizlabs.android.dbflow.processor.ClassNames
import com.raizlabs.android.dbflow.processor.ModelViewValidator
import com.raizlabs.android.dbflow.processor.ProcessorManager
import com.raizlabs.android.dbflow.processor.TableValidator
import com.raizlabs.android.dbflow.processor.utils.`override fun`
import com.raizlabs.android.dbflow.processor.utils.annotation
import com.squareup.javapoet.ClassName
import com.squareup.javapoet.ParameterizedTypeName
import com.squareup.javapoet.TypeName
import com.squareup.javapoet.TypeSpec
import com.squareup.javapoet.WildcardTypeName
import java.util.*
import java.util.regex.Pattern
import javax.lang.model.element.Element

/**
 * Description: Writes [Database] definitions,
 * which contain [Table], [ModelView], and [Migration]
 */
class DatabaseDefinition(manager: ProcessorManager, element: Element) : BaseDefinition(element, manager), TypeDefinition {

    var databaseClassName: String? = null

    var databaseVersion: Int = 0

    internal var foreignKeysSupported: Boolean = false

    internal var consistencyChecksEnabled: Boolean = false

    internal var backupEnabled: Boolean = false

    var insertConflict: ConflictAction? = null

    var updateConflict: ConflictAction? = null

    var classSeparator: String = ""
    var fieldRefSeparator: String = "" // safe field for javapoet

    var objectHolder: DatabaseObjectHolder? = null

    var databaseExtensionName = ""

    init {
        packageName = ClassNames.FLOW_MANAGER_PACKAGE

        element.annotation<Database>()?.let { database ->
            databaseExtensionName = database.databaseExtension
            databaseClassName = element.simpleName.toString()
            consistencyChecksEnabled = database.consistencyCheckEnabled
            backupEnabled = database.backupEnabled

            classSeparator = database.generatedClassSeparator
            fieldRefSeparator = classSeparator

            setOutputClassName(databaseClassName + classSeparator + "Database")

            databaseVersion = database.version
            foreignKeysSupported = database.foreignKeyConstraintsEnforced

            insertConflict = database.insertConflict
            updateConflict = database.updateConflict
        }
    }

    override val extendsClass: TypeName? = ClassNames.BASE_DATABASE_DEFINITION_CLASSNAME

    override fun onWriteDefinition(typeBuilder: TypeSpec.Builder) {
        writeConstructor(typeBuilder)
        writeGetters(typeBuilder)
    }

    fun validateAndPrepareToWrite() {
        prepareDefinitions()
        validateDefinitions()
    }

    private fun validateDefinitions() {
        elementClassName?.let {
            val map = HashMap<TypeName, TableDefinition>()
            val tableValidator = TableValidator()
            manager.getTableDefinitions(it)
                .filter { tableValidator.validate(ProcessorManager.manager, it) }
                .forEach { it.elementClassName?.let { className -> map.put(className, it) } }
            manager.setTableDefinitions(map, it)

            val modelViewDefinitionMap = HashMap<TypeName, ModelViewDefinition>()
            val modelViewValidator = ModelViewValidator()
            manager.getModelViewDefinitions(it)
                .filter { modelViewValidator.validate(ProcessorManager.manager, it) }
                .forEach { it.elementClassName?.let { className -> modelViewDefinitionMap.put(className, it) } }
            manager.setModelViewDefinitions(modelViewDefinitionMap, it)
        }
    }

    private fun prepareDefinitions() {
        elementClassName?.let {
            manager.getTableDefinitions(it).forEach(TableDefinition::prepareForWrite)
            manager.getModelViewDefinitions(it).forEach(ModelViewDefinition::prepareForWrite)
            manager.getQueryModelDefinitions(it).forEach(QueryModelDefinition::prepareForWrite)
        }
    }

    private fun writeConstructor(builder: TypeSpec.Builder) {

        builder.constructor(param(ClassNames.DATABASE_HOLDER, "holder")) {
            modifiers(public)
            this@DatabaseDefinition.elementClassName?.let { elementClassName ->
                for (definition in manager.getTableDefinitions(elementClassName)) {
                    if (definition.hasGlobalTypeConverters) {
                        statement("addModelAdapter(new \$T(holder, this), holder)", definition.outputClassName)
                    } else {
                        statement("addModelAdapter(new \$T(this), holder)", definition.outputClassName)
                    }
                }

                for (definition in manager.getModelViewDefinitions(elementClassName)) {
                    if (definition.hasGlobalTypeConverters) {
                        statement("addModelViewAdapter(new \$T(holder, this), holder)", definition.outputClassName)
                    } else {
                        statement("addModelViewAdapter(new \$T(this), holder)", definition.outputClassName)
                    }
                }

                for (definition in manager.getQueryModelDefinitions(elementClassName)) {
                    if (definition.hasGlobalTypeConverters) {
                        statement("addQueryModelAdapter(new \$T(holder, this), holder)", definition.outputClassName)
                    } else {
                        statement("addQueryModelAdapter(new \$T(this), holder)", definition.outputClassName)
                    }
                }

                val migrationDefinitionMap = manager.getMigrationsForDatabase(elementClassName)
                migrationDefinitionMap.keys
                    .sortedByDescending { it }
                    .forEach { version ->
                        migrationDefinitionMap[version]
                            ?.sortedBy { it.priority }
                            ?.forEach { migrationDefinition ->
                                statement("addMigration($version, new \$T${migrationDefinition.constructorName})", migrationDefinition.elementClassName)
                            }
                    }
            }
            this
        }

    }

    private fun writeGetters(typeBuilder: TypeSpec.Builder) {
        typeBuilder.apply {
            `override fun`(ParameterizedTypeName.get(ClassName.get(Class::class.java), WildcardTypeName.subtypeOf(Any::class.java)),
                "getAssociatedDatabaseClassFile") {
                modifiers(public, final)
                `return`("\$T.class", elementTypeName)
            }
            `override fun`(TypeName.BOOLEAN, "isForeignKeysSupported") {
                modifiers(public, final)
                `return`(foreignKeysSupported.L)
            }
            `override fun`(TypeName.BOOLEAN, "backupEnabled") {
                modifiers(public, final)
                `return`(backupEnabled.L)
            }
            `override fun`(TypeName.BOOLEAN, "areConsistencyChecksEnabled") {
                modifiers(public, final)
                `return`(consistencyChecksEnabled.L)
            }
            `override fun`(TypeName.INT, "getDatabaseVersion") {
                modifiers(public, final)
                `return`(databaseVersion.L)
            }
            if (!databaseExtensionName.isNullOrBlank()) {
                `override fun`(String::class, "getDatabaseExtensionName") {
                    modifiers(public, final)
                    `return`(databaseExtensionName.S)
                }
            }
        }
    }

    /**
     *
     * Checks if databaseName is valid. It will check if databaseName matches regex pattern
     * [A-Za-z_$]+[a-zA-Z0-9_$]* Examples:   * database - valid  * DbFlow1 - valid  * database.db -
     * invalid (contains a dot)  * 1database - invalid (starts with a number)

     * @param databaseName database name to validate.
     * *
     * @return `true` if parameter is a valid database name, `false` otherwise.
     */
    private fun isValidDatabaseName(databaseName: String?): Boolean {
        val javaClassNamePattern = Pattern.compile("[A-Za-z_$]+[a-zA-Z0-9_$]*")
        return javaClassNamePattern.matcher(databaseName).matches()
    }
}
